tasks/nunjucks.js

```diff
@@ -27,112 +59,67 @@ module.exports = function (grunt) {
-          grunt.log.error(`No source files specified for ${chalk.cyan(filedest)}.`)
-
-          // Skip to next file — nothing we can do without specified source files
-          return reject(new Error('For some destinations were not specified source files.'))
-        }
-
-        // Iterate over files' sources
-        file.src.forEach(src => {
-          // Construct absolute path to file for Nunjucks
-          let filepath = path.join(process.cwd(), src)
-
-          let data = {}
-          // Clone data
-          for (let i in options.data) {
-            if (options.data.hasOwnProperty(i)) {
-              data[i] = options.data[i]
-            }
-          }
-
-          // Preprocess data
-          if (options.data && typeof options.preprocessData === 'function') {
-            data = options.preprocessData.call(file, data)
-          }
-
-          // Asynchronously render templates with configurated Nunjucks environment
-          // and write to destination
-          env.render(filepath, data, (error, result) => {
-            // Catch errors, warn
-            if (error) {
-              grunt.log.error(error)
-              grunt.fail.warn('Failed to compile one of the source files.')
-              grunt.log.writeln()
-
-              // Prevent writing of failed to compile file, skip to next file
-              return reject(new Error('Failed to compile some source files.'))
-            }
-
-            // Write rendered template to destination
-            grunt.file.write(filedest, result)
-
-            // Debug process
-            grunt.verbose.ok(`File ${chalk.cyan(filedest)} created.`)
-            grunt.verbose.writeln()
-
-            countCompiled++
-          })
-        })
-      })
+    let success = 0
+    let errors = 0
 
-      // Finish Promise
-      resolve()
-    })
+    const renderFiles = this.files.reduce((files, file) => {
+      let filedest = file.dest
 
-    // Print any errors from rejects
-    task.catch(error => {
-      if (error) {
-        grunt.log.writeln()
-        grunt.log.error(error)
-        grunt.log.writeln()
+      if (!filedest) {
+        grunt.log.error(`No dest file specified`)
+        done()
       }
-    })
 
-    // Log number of processed templates
-    task.then(success => {
-      // Log number of processed templates
-      let logType = (countCompiled === totalFiles) ? 'ok' : 'error'
-      let countCompiledColor = (logType === 'ok') ? 'green' : 'red'
-      grunt.log[logType](`${chalk[countCompiledColor](countCompiled)}/${chalk.cyan(totalFiles)} ${grunt.util.pluralize(totalFiles, 'file/files')} compiled.`)
-    })
+      if (!file.src || !file.src.length) {
+        grunt.log.error(`No source files specified for ${chalk.cyan(filedest)}.`)
+        done()
+      }
 
-    // Finish async task
-    completeTask()
+      const renders = file.src.map((src) => new Promise((resolve, reject) =>
+        render(env, path.resolve(src), prepareData(options, file))
```

Comment:
`render` returns a promise, why do we need another one above?

Probably it should be:

```js
const renders = file.src.map((scr) => {
  return render(
    env, path.resolve(src), prepareData(options, file)
  )
    .then(...)
    .catch(...)
    .then(...)
})
```

?

Reply:
Ah, great catch! After all those years, I still able to mess up Promises :) I was wondering all the day long why I need to call that trailing resolve, and that it looks odd. Indeed, totally unnecessary Promise.

I've pushed a fix.